validate.swift

```diff
@@ -2,67 +2,450 @@
-// Find the "packages.json" file based on arguments, current directory, or the directory of the script
-let argumentURL = CommandLine.arguments.dropFirst().first.flatMap(URL.init(fileURLWithPath:))
-let currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath, isDirectory: true).appendingPathComponent("packages.json")
-let scriptDirectoryURL = URL(fileURLWithPath: #file).deletingLastPathComponent().appendingPathComponent("packages.json")
+// MARK: Configuration Values and Constants
+
+// number of validations to run simultaneously
+let semaphoreCount = 12
+
+let timeoutIntervalForRequest = 3000.0
+let timeoutIntervalForResource = 6000.0
+
+// base url for github raw files
+let rawURLComponentsBase = URLComponents(string: "https://raw.githubusercontent.com")!
+
+// master package list to compare against
+let masterPackageList = rawURLComponentsBase.url!.appendingPathComponent("daveverwer/SwiftPMLibrary/master/packages.json")
 
-let possibleURLs : [URL?] = [argumentURL, currentDirectoryURL, scriptDirectoryURL]
+let helpText = """
+usage: %@ <command> [path]
 
-guard let url = possibleURLs.compactMap({ $0 }).first( where: { FileManager.default.fileExists(atPath: $0.path )}) else {
-  print("Error: Unable to find packages.json to validate.")
-  exit(1)
+COMMANDS:
+  all   validate all packages in JSON packages.json
+  diff  validate all new packages in JSON packages.json
+  mine  validate the Package of the current directoy
+
+OPTIONS:
+  path  to define the specific `packages.json` file or Swift package directory
+"""
+// MARK: Types
+
+
+enum Command : String {
+  case all = "all"
+  case diff = "diff"
+  case mine = "mine"
 }
 
-let decoder = JSONDecoder()
-let data = try! Data(contentsOf: url)
-let packageUrls  = try! decoder.decode([URL].self, from: data)
+extension Command {
+  static func fromArguments (_ arguments : [String]) -> Command? {
+    for argument in arguments {
+      if let command = Command(rawValue: argument) {
+        return command
+      }
+    }
+    return nil
+  }
+}
 
-// Make sure all urls contain the .git extension
-print("Checking all urls are valid.")
-let invalidUrls = packageUrls.filter{ $0.pathExtension != "git" }
+/**
+ Simple Product structure from package dump
+ */
+struct Product: Codable {
+  let name: String
+}
 
-guard invalidUrls.count == 0 else {
-  print("Invalid URLs missing .git extension: \(invalidUrls)")
-  exit(1)
+/**
+ Simple Package structure from package dump
+ */
+struct Package: Codable {
+  let name: String
+  let products: [Product]
 }
 
-// Make sure there are no dupes (no dupe variants w/ .git and w/o, no case differences)
-print("Checking for duplicate packages.")
-let urlCounts = Dictionary(grouping: packageUrls.enumerated()) {
-  URL(string: $0.element.absoluteString.lowercased())!
-}.mapValues{ $0.map{ $0.offset }}.filter{$0.value.count > 1}
+/**
+ List of git hosts for which we can pull single files
+ */
+enum GitHost: String {
+  case GitHub = "github.com"
```

Comment:
Couldn't that be just a local variable?

Reply:
This is for future-proofing support for BitBucket, GitLab, etc... 
For content check out my comment [here](https://github.com/daveverwer/SwiftPMLibrary/issues/215#issuecomment-597688250)